File: NetFramework/XHtmlEditor/CS/C1WordsX/Tabs/ViewTab.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Windows.Forms;
using System.Drawing;
using C1.Win.C1Ribbon;

using EditorMode = C1.Win.C1Editor.EditorMode;

namespace C1WordsX
{
    class ViewTab : C1TextEditorRibbonTab
    {
        #region declarations

        private RibbonToggleButton _modeDesign;
        private RibbonToggleButton _modeSource;
        private RibbonToggleButton _modePreview;
        private RibbonButton _fixErrors;
        private RibbonButton _validateDocument;
        private bool _isFixing;

        #endregion

        #region ctors

        public ViewTab()
        {
            this.ID = this.Text = "View";
            Groups.Add(CreateDocumentViewsGroup());
            Groups.Add(CreateSourceGroup());
            _isFixing = false;
        }

        #endregion

        #region overrides

        internal override void InitTab()
        {
            Ribbon.EditorTextChanged += EditorTextChangedHandler;
            base.InitTab();
        }

        protected override void OnSelect(EventArgs e)
        {
            base.OnSelect(e);
            UpdateEditorModeButtons();
        }

        internal override void HandleItemEvent(RibbonEventArgs e)
        {
            RibbonItem item = (RibbonItem)e.Item;
            if (e.EventType == RibbonEventType.Click)
            {
                if (item == _modeDesign)
                    SetEditorMode(EditorMode.Design);
                else if (item == _modeSource)
                    SetEditorMode(EditorMode.Source);
                else if (item == _modePreview)
                    SetEditorMode(EditorMode.Preview);
                else if (item == _fixErrors)
                {
                    _isFixing = true;
                    try
                    {
                        Editor.FixSource();
                    }
                    finally
                    {
                        _isFixing = false;
                    }
                    _fixErrors.Enabled = false;
                    _validateDocument.Enabled = false;
                }
                else if (item == _validateDocument)
                {
                    bool errorsFixed = (Editor as C1RibbonEditorXhtml).ValidateSource();
                    _fixErrors.Enabled = !errorsFixed;
                    _validateDocument.Enabled = false;
                }
            }
        }

        #endregion

        #region implementation

        private void UpdateEditorModeButtons()
        {
            C1RibbonEditorXhtml ed = Editor as C1RibbonEditorXhtml;

            _modeDesign.Visible = ed != null;
            _modeSource.Visible = ed != null;
            _modePreview.Visible = ed != null;
            if (ed != null)
            {
                _modeDesign.Pressed = ed.Mode == EditorMode.Design;
                _modeSource.Pressed = ed.Mode == EditorMode.Source;
                _modePreview.Pressed = ed.Mode == EditorMode.Preview;
                Groups["Validation"].Visible = ed.Mode == EditorMode.Source;
            }
        }

        private void UpdateRibbonState(EditorMode mode)
        {
            RibbonTab home = Ribbon.Tabs["Home"];
            home.Groups["Clipboard"].Enabled = mode != EditorMode.Preview;
            home.Groups["Font"].Enabled = mode == EditorMode.Design;
            home.Groups["Paragraph"].Enabled = mode == EditorMode.Design;
            home.Groups["Editing"].Enabled = mode != EditorMode.Preview;
            RibbonTab insert = Ribbon.Tabs["Insert"];
            insert.Groups["Tables"].Enabled = mode == EditorMode.Design;
            insert.Groups["Illustrations"].Enabled = mode == EditorMode.Design;
            insert.Groups["Links"].Enabled = mode == EditorMode.Design;
            insert.Groups["Text"].Enabled = mode == EditorMode.Design;
            RibbonTab review = Ribbon.Tabs["Review"];
            review.Groups["Proofing"].Enabled = mode == EditorMode.Design;            
        }

        private RibbonGroup CreateDocumentViewsGroup()
        {
            RibbonGroup g = new RibbonGroup();
            g.Text = g.ID = "Document Views";

            g.Items.Add(_modeDesign = CreateToggleButton("EditorDesignMode"));
            g.Items.Add(_modeSource = CreateToggleButton("EditorSourceMode"));
            g.Items.Add(_modePreview = CreateToggleButton("EditorPreviewMode"));

            return g;
        }

        private RibbonGroup CreateSourceGroup()
        {
            RibbonGroup g = new RibbonGroup();
            g.Text = g.ID = "Validation";

            g.Items.Add(_fixErrors = CreateButton("FixErrors"));
            g.Items.Add(_validateDocument = CreateButton("ValidateDocument"));

            return g;
        }

        private void SetEditorMode(EditorMode mode)
        {
            C1RibbonEditorXhtml ed = Editor as C1RibbonEditorXhtml;
            ed.Mode = mode;
            _fixErrors.Enabled = false;
            _validateDocument.Enabled = false;
            UpdateEditorModeButtons();
            if (ed.Mode == mode)
                UpdateRibbonState(mode);
            if (ed.Mode != mode && ed.Mode == EditorMode.Source)
            {
                _fixErrors.Enabled = true;
                _validateDocument.Enabled = false;
            }
        }

        private void EditorTextChangedHandler(object sender, EventArgs e)
        {
            C1RibbonEditorXhtml ed = Editor as C1RibbonEditorXhtml;
            if (ed != null && ed.Mode == EditorMode.Source && !_isFixing)
            {
                _validateDocument.Enabled = true;
            }
        }

        #endregion
    }
}

File: NetFramework/XHtmlEditor/VB/RichTextEditor/Tabs/ViewTab.vb
Imports System.Collections.Generic
Imports System.Text
Imports System.Windows.Forms
Imports System.Drawing
Imports C1.Win.Ribbon

Imports EditorMode = C1.Win.C1Editor.EditorMode

Namespace RichTextEditor
	Class ViewTab
		Inherits C1TextEditorRibbonTab
		#Region "declarations"

		Private _modeDesign As RibbonToggleButton
		Private _modeSource As RibbonToggleButton
		Private _modePreview As RibbonToggleButton
		Private _fixErrors As RibbonButton
		Private _validateDocument As RibbonButton
		Private _isFixing As Boolean

		#End Region

		#Region "ctors"

		Public Sub New()
			Me.ID = InlineAssignHelper(Me.Text, "View")
			Groups.Add(CreateDocumentViewsGroup())
			Groups.Add(CreateSourceGroup())
			_isFixing = False
		End Sub

		#End Region

		#Region "overrides"

		Friend Overloads Overrides Sub InitTab()
			AddHandler Ribbon.EditorTextChanged, AddressOf EditorTextChangedHandler
			MyBase.InitTab()
		End Sub

		Protected Overloads Overrides Sub OnSelect(e As EventArgs)
			MyBase.OnSelect(e)
			UpdateEditorModeButtons()
		End Sub

		Friend Overloads Overrides Sub HandleItemEvent(e As RibbonEventArgs)
			Dim item As RibbonItem = DirectCast(e.Item, RibbonItem)
			If e.EventType = RibbonEventType.Click Then
				If item Is _modeDesign Then
					SetEditorMode(EditorMode.Design)
				ElseIf item Is _modeSource Then
					SetEditorMode(EditorMode.Source)
				ElseIf item Is _modePreview Then
					SetEditorMode(EditorMode.Preview)
				ElseIf item Is _fixErrors Then
					_isFixing = True
					Try
						Editor.FixSource()
					Finally
						_isFixing = False
					End Try
					_fixErrors.Enabled = False
					_validateDocument.Enabled = False
				ElseIf item Is _validateDocument Then
					Dim errorsFixed As Boolean = TryCast(Editor, C1RibbonEditorXhtml).ValidateSource()
					_fixErrors.Enabled = Not errorsFixed
					_validateDocument.Enabled = False
				End If
			End If
		End Sub

		#End Region

		#Region "implementation"

		Private Sub UpdateEditorModeButtons()
			Dim ed As C1RibbonEditorXhtml = TryCast(Editor, C1RibbonEditorXhtml)

			_modeDesign.Visible = ed IsNot Nothing
			_modeSource.Visible = ed IsNot Nothing
			_modePreview.Visible = ed IsNot Nothing
			If ed IsNot Nothing Then
				_modeDesign.Pressed = ed.Mode = EditorMode.Design
				_modeSource.Pressed = ed.Mode = EditorMode.Source
				_modePreview.Pressed = ed.Mode = EditorMode.Preview
				Groups("Validation").Visible = ed.Mode = EditorMode.Source
			End If
		End Sub

		Private Sub UpdateRibbonState(mode As EditorMode)
			Dim home As RibbonTab = Ribbon.Tabs("Home")
			home.Groups("Clipboard").Enabled = mode <> EditorMode.Preview
			home.Groups("Font").Enabled = mode = EditorMode.Design
			home.Groups("Paragraph").Enabled = mode = EditorMode.Design
            home.Groups("Editing").Enabled = mode <> EditorMode.Preview
			Dim insert As RibbonTab = Ribbon.Tabs("Insert")
			insert.Groups("Tables").Enabled = mode = EditorMode.Design
			insert.Groups("Illustrations").Enabled = mode = EditorMode.Design
			insert.Groups("Links").Enabled = mode = EditorMode.Design
			insert.Groups("Text").Enabled = mode = EditorMode.Design
			Dim review As RibbonTab = Ribbon.Tabs("Review")
			review.Groups("Proofing").Enabled = mode = EditorMode.Design
		End Sub

		Private Function CreateDocumentViewsGroup() As RibbonGroup
			Dim g As RibbonGroup = New RibbonGroup()
			g.Text = InlineAssignHelper(g.ID, "Document Views")

			g.Items.Add(InlineAssignHelper(_modeDesign, CreateToggleButton("EditorDesignMode")))
			g.Items.Add(InlineAssignHelper(_modeSource, CreateToggleButton("EditorSourceMode")))
			g.Items.Add(InlineAssignHelper(_modePreview, CreateToggleButton("EditorPreviewMode")))

			Return g
		End Function

		Private Function CreateSourceGroup() As RibbonGroup
			Dim g As RibbonGroup = New RibbonGroup()
			g.Text = InlineAssignHelper(g.ID, "Validation")

			g.Items.Add(InlineAssignHelper(_fixErrors, CreateButton("FixErrors")))
			g.Items.Add(InlineAssignHelper(_validateDocument, CreateButton("ValidateDocument")))

			Return g
		End Function

		Private Sub SetEditorMode(mode As EditorMode)
			Dim ed As C1RibbonEditorXhtml = TryCast(Editor, C1RibbonEditorXhtml)
			ed.Mode = mode
			_fixErrors.Enabled = False
			_validateDocument.Enabled = False
			UpdateEditorModeButtons()
			If ed.Mode = mode Then
				UpdateRibbonState(mode)
			End If
			If ed.Mode <> mode AndAlso ed.Mode = EditorMode.Source Then
				_fixErrors.Enabled = True
				_validateDocument.Enabled = False
			End If
		End Sub

		Private Sub EditorTextChangedHandler(sender As Object, e As EventArgs)
			Dim ed As C1RibbonEditorXhtml = TryCast(Editor, C1RibbonEditorXhtml)
			If ed IsNot Nothing AndAlso ed.Mode = EditorMode.Source AndAlso Not _isFixing Then
				_validateDocument.Enabled = True
			End If
		End Sub
		Private Shared Function InlineAssignHelper(Of T)(ByRef target As T, value As T) As T
			target = value
			Return value
		End Function

		#End Region
	End Class
End Namespace
